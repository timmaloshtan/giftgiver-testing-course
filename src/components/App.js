import React, { Component } from 'react';
import { Button } from 'react-bootstrap'

import Gift from './Gift';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      gifts: []
    };
  }

  addGift = () => {
    this.setState(prevState => ({
      gifts: [
        ...prevState.gifts,
        {
          id: prevState.gifts.length + 1
        },
      ],
    }));
  }

  removeGift = (id) => {
    const { gifts } = this.state;

    this.setState({
      gifts: gifts.filter(gift => gift.id !== id)
    });
  }

  render() {
    return (
      <div className="app">
        <h2>Gift Giver</h2>
        <div className="gift-list">
          {
            this.state.gifts.map(gift => (
              <Gift
                key={gift.id}
                gift={gift}
                removeGift={this.removeGift}
              />
            ))
          }
        </div>
        <Button
          className="btn-add"
          onClick={this.addGift}
        >
          Add Gift
        </Button>
      </div>
    );
  }
}

export default App;