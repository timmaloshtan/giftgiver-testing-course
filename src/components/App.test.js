import React from 'react';
import { shallow } from 'enzyme'

import App from './App.js';

describe('App component', () => {
  const app = shallow(<App />);
  
  it('should render correctly', () => {
    expect(app).toMatchSnapshot();
  });

  it('should initialize the state with an empty list of gifts', () => {
    expect(app.state().gifts).toEqual([]);
  });

  describe('when clicking the "add gift" button', () => {
    const GIFT_ID = 1;

    beforeEach(() => {
      app.find('.btn-add').simulate('click');
    });
    afterEach(() => {
      app.setState({ gifts: [] });
    });

    it('should add a new gift to state', () => {
      expect(app.state().gifts).toEqual([{ id: GIFT_ID }]);
    });

    it('should add a new gift to gift list', () => {
      expect(app.find('.gift-list').children().length).toEqual(1);
    });

    it('should create a new Gift component', () => {
      expect(app.find('Gift').exists()).toBe(true);
    });

    describe('and removind the added gift', () => {
      beforeEach(() => {
        app.instance().removeGift(GIFT_ID);
      });

      it('should remove the gift from gift-list', () => {
        expect(app.find('Gift').exists()).toBe(false);
      });
    });
  });
});
