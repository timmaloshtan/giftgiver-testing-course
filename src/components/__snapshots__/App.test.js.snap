// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render correctly 1`] = `
<div
  className="app"
>
  <h2>
    Gift Giver
  </h2>
  <div
    className="gift-list"
  />
  <Button
    active={false}
    className="btn-add"
    disabled={false}
    onClick={[Function]}
    type="button"
    variant="primary"
  >
    Add Gift
  </Button>
</div>
`;
